src/main/java/com/cloudbees/hudson/plugins/folder/ChildNameGenerator.java

```diff
@@ -0,0 +1,290 @@
+     *     <li>Because of Windows and allowing for users to migrate their Jenkins from unix to windows and vice-versa,
+     *     some names are reserved names under Windows:
+     *     {@code AUX, COM1, COM2, ..., COM9, CON, LPT1, LPT2, ..., LPT9, NUL, PRN} plus all case variations of these
+     *     names plus the variants where a single {@code .} is appended, you need to map those to something else</li>
+     *     <li>Don't make the filenames too long. Try to keep them under 32 characters. If you can go smaller, even
+     *     better.</li>
+     *     <li>Get it right first time</li>
+     * </ul>
+     *
+     * @param parent the parent within which the item is being loaded.
+     * @param item   the partially loaded item (take care what methods you call, the item will not have a reference to
+     *               its parent.
+     * @return the filesystem safe mangled equivalent name of the item.
+     */
+    @CheckForNull
+    public abstract String dirNameFromItem(@Nonnull P parent, @Nonnull I item);
+
+    /**
+     * {@link #itemNameFromItem(AbstractFolder, TopLevelItem)} could not help, we are loading the item for the first
+     * time since the {@link ChildNameGenerator} was enabled for the parent folder type, this method's mission is
+     * to pretend the {@code legacyDirName} is the "mostly correct" name and turn this into the actual name.
+     *
+     * Challenges include:
+     * <ul>
+     * <li>Previously the name may have been over-encoded with {@link Util#rawEncode(String)} so you may need to
+     * decode it first</li>
+     * <li>There are some characters that it would be really bad to return in the item name, such as
+     * {@code "/" / "?" / "#" / "[" / "]" / "\"} as these could end up modifying the effective URL</li>
+     * <li>There are names that it would be bad to return as the item name, such as
+     * {@code "" / "." / ".."} as these could end creating broken effective URLs</li>
+     * </ul>
+     * @param parent the parent within which the item is being loaded.
+     * @param legacyDirName the directory name that we are loading an item from.
+     * @return
+     */
+    @Nonnull
+    public abstract String itemNameFromLegacy(@Nonnull P parent, @Nonnull String legacyDirName);
+
+    /**
+     * {@link #dirNameFromLegacy(AbstractFolder, String)} could not help, we are loading the item for the first
+     * time since the {@link ChildNameGenerator} was enabled for the parent folder type, this method's mission is
+     * to pretend the {@code legacyDirName} is the "mostly correct" name and turn this into the filesystem safe
+     * mangled equivalent name to use going forward.
+     *
+     * Challenges include:
+     * <ul>
+     *     <li>The only really filesystem safe characters are {@code A-Za-z0-9_.-}</li>
+     *     <li>Because of Windows and allowing for users to migrate their Jenkins from unix to windows and vice-versa,
```

Comment:
Typos. And heavily duplicated documentation. Move it to class Javadoc and refer to that from here.